Std/Tactic/LibrarySearch.lean

```diff
@@ -0,0 +1,531 @@
+
+open Lean Meta Std.Tactic.TryThis
+
+initialize registerTraceClass `Tactic.stdLibrarySearch
+initialize registerTraceClass `Tactic.stdLibrarySearch.lemmas
+
+/-- Configuration for `DiscrTree`. -/
+def discrTreeConfig : WhnfCoreConfig := {}
+
+/--
+A "modifier" for a declaration.
+* `none` indicates the original declaration,
+* `mp` indicates that (possibly after binders) the declaration is an `↔`,
+  and we want to consider the forward direction,
+* `mpr` similarly, but for the backward direction.
+-/
+inductive DeclMod
+  | /-- the original declaration -/ none
+  | /-- the forward direction of an `iff` -/ mp
+  | /-- the backward direction of an `iff` -/ mpr
+deriving DecidableEq, Inhabited, Ord
+
+instance : ToString DeclMod where
+  toString m := match m with | .none => "" | .mp => "mp" | .mpr => "mpr"
+
+/--
+LibrarySearch has an extension mechanism for replacing the function used
+to find candidate lemmas.
+-/
+@[reducible]
+def CandidateFinder := Expr → MetaM (Array (Name × DeclMod))
+
+open LazyDiscrTree (InitEntry isBlackListed createImportedEnvironment)
+
+namespace DiscrTreeFinder
+
+open System (FilePath)
+
+
+/--
+Once we reach Mathlib, and have `cache` available,
+we may still want to load a precomputed cache for `exact?` from a `.olean` file.
+
+This makes no sense here in Std, where there is no caching mechanism.
+-/
+def cachePath : IO FilePath := do
+  let sp ← searchPathRef.get
+  if let buildPath :: _ := sp then
+    let path := buildPath / "LibrarySearch.extra"
+    if ← path.pathExists then
+      return path
+  return ".lake" / "build" / "lib" / "LibrarySearch.extra"
+
+/-- Add a path to a discrimination tree.-/
+private def addPath [BEq α] (config : WhnfCoreConfig) (tree : DiscrTree α) (tp : Expr) (v : α) :
+    MetaM (DiscrTree α) := do
+  let k ← DiscrTree.mkPath tp config
+  pure <| tree.insertCore k v config
+
+/-- Adds a constant with given name to tree. -/
+private def updateTree (config : WhnfCoreConfig) (tree : DiscrTree (Name × DeclMod))
+    (name : Name) (constInfo : ConstantInfo) : MetaM (DiscrTree (Name × DeclMod)) := do
+  if constInfo.isUnsafe then return tree
+  if isBlackListed (←getEnv) name then return tree
+  withReducible do
+    let (_, _, type) ← forallMetaTelescope constInfo.type
+    let tree ← addPath config tree type (name, DeclMod.none)
+    match type.getAppFnArgs with
+    | (``Iff, #[lhs, rhs]) => do
+      let tree ← addPath config tree rhs (name, DeclMod.mp)
+      let tree ← addPath config tree lhs (name, DeclMod.mpr)
+      return tree
+    | _ =>
+      return tree
+
+/--
+Constructs an discriminator tree from the current environment.
+-/
+def buildImportCache (config : WhnfCoreConfig) : MetaM (DiscrTree (Name × DeclMod)) := do
+  let profilingName := "apply?: init cache"
+  -- Sort so lemmas with longest names come first.
+  let post (A : Array (Name × DeclMod)) :=
+        A.map (fun (n, m) => (n.toString.length, n, m)) |>.qsort (fun p q => p.1 > q.1) |>.map (·.2)
```

Comment:
There's no analogue of this in the no-cache-file version, right?

It's not essential, but this does seem to be a genuinely helpful way of organising results. We can revisit later.

Reply:
I don't like using the names for ordering as a general principle.

The lazy discrimination tree orders results in reverse order they are found which is an approximation of match size with patterns in the same location being ordered in the order they are imported.

By "match size" I mean the number of keys other than star that are shared between the term and pattern.

I'll see if I can use a heap to return actual largest match first with import order for ties.

Reply:
I added the ability to use longest match first.  We can adjust as needed in response to feedback.